src/NHibernate.Test/CacheTest/QueryCacheFixture.cs

```diff
@@ -60,5 +60,19 @@ public void QueryCacheWithNullParameters()
 				s.Flush();
 			}
 		}
+
+		[Test]
+		public void QueryCacheWithScalarFunctionThatReturnsInt()
+		{
+			using (ISession s = OpenSession())
+			{
+				int result = s
+					.CreateSQLQuery("SELECT 200012")
+					.SetCacheable(true)
+					.UniqueResult<int>();
+
+				Assert.IsNotNull(result);
```

Comment:
This test is not failing even without your fixes. It needs to run two times the query for validating the fix.

Reply:
In fact it fails without your changes, but with your changes trying then to re-query (so trying to retrieve the result from the cache) then fails.